File: highlight.lua
--- Simple Syntax Highlighting

local M = {}

--- Wrap corresponding parts of Lua-Code with HTML-Spans of the classes
-- 'comment', 'string', 'escape', 'operator' and 'keyword'. Does not really
-- work with multiline comments or multiline strings, everything else should
-- be fine.
--[=[
local highlight = require "highlight"

return highlight.highlight( [[
   function get(table, key)
       return table[key]
   end
]] )
]=]--@RUN
function M.highlight(string)
    local inserts = {}

    local function wrapclass(class) return function(x, y)
        if not y then y = "" end
        table.insert(inserts, "<span class=" .. class .. ">" .. x .. "</span>")
        return "$" .. #inserts .. ";" .. y
    end end

    local function suspend(x)
        table.insert(inserts, x)
        return "$" .. #inserts .. ";"
    end

    return (string
        :gsub("<", "&lt;") .. "<")
        :gsub("\n", "<br>")

        :gsub("%$", suspend)
--        :gsub("%$.-;", suspend)

        :gsub("\\\\", suspend)
        :gsub('\\["\']', wrapclass("escape"))
--        :gsub("(<br>%-%-.-)([\n<])", wrapclass("comment"))
        :gsub("'.-'", wrapclass("string")):gsub('".-"', wrapclass("string"))
        :gsub("(%-%-.-)([\n<])", wrapclass("comment"))

--        :gsub("(%d+)([^;])", wrapclass("number"))

        :gsub("[=()#[%]+-*/{}]", wrapclass("operator"))

        :gsub("(elseif)([ \n<])", wrapclass("keyword"))
        :gsub("(do)([ \n<])", wrapclass("keyword"))
        :gsub("(in)([ \n<])", wrapclass("keyword"))
        :gsub("(if)([ \n<])", wrapclass("keyword"))
        :gsub("(for)([ \n<])", wrapclass("keyword"))
        :gsub("(end)([ \n<])", wrapclass("keyword"))
        :gsub("(else)([ \n<])", wrapclass("keyword"))
        :gsub("(then)([ \n<])", wrapclass("keyword"))
        :gsub("(break)([ \n<])", wrapclass("keyword"))
        :gsub("(until)([ \n<])", wrapclass("keyword"))
        :gsub("(local)([ \n<])", wrapclass("keyword"))
        :gsub("(while)([ \n<])", wrapclass("keyword"))
        :gsub("(return)([ \n<])", wrapclass("keyword"))
        :gsub("(function)([ \n<])", wrapclass("keyword"))


        :gsub(" ", "&nbsp;")

        :gsub("$(.-);", function (x) return inserts[tonumber(x)] end)
        :gsub("$(.-);", function (x) return inserts[tonumber(x)] end)

        :sub(1, -2)

-- debugging
--        .. "</pre>" .. doc.wraps(doc.map(inserts, function (x) return "------<br>" .. x end ), "p")
end

return M

File: highlight.js

function highlight(string) {
    "use strict";

    var inserts = {};
    var i = 0;

    var wrapclass = function(klass) { return function (_, x, y) {
        inserts[i] = "<span class=" + klass + ">" + x + "</span>";
        return "$" + i++ + ";" + y;
    }};

    var suspend = function(x) {
        inserts[i] = ""+x;
        return "$" + i++ + ";";
    };

    var plug = function(x) {
        var v = inserts[parseInt(x.slice(1,-1))];
        return v===undefined? x: v;
    };
    var plug2 = function(x) {
        var v = inserts[parseInt(x.slice(1,-1))];
        return v==="$" || v===undefined? x: v;
    };

    return (string .replace(/&/g, "&amp;") .replace(/</g, "&lt;") + "<br>")

        .replace(/\n/g, "<br>")
        .replace(/\t/g, "    ")
        .replace(/ /g, "<wbr>&nbsp;")

        .replace(/\$/g, suspend)

        .replace(/(\\\\)()/g, wrapclass("escape"))
        .replace(/(\\["'])()/g, wrapclass("escape"))

        .replace(/('.+?')()/g, wrapclass("string"))
        .replace(/(".+?")()/g, wrapclass("string"))

        .replace(/(--.+?)(<br>)/g, wrapclass("comment"))

        /*.replace(/(d+)([^;])/g, wrapclass("number"))*/
        .replace(/([=()#[\]+*\-{}])()/g, wrapclass("operator"))

        .replace(/(elseif|do|in|if|for|end|else|then|break|until|local|while|return|function)(<)/g, wrapclass("keyword"))

        .replace(/\$.+?;/g, plug2)
        .replace(/\$.+?;/g, plug2)
        .replace(/\$.+?;/g, plug)

        .slice(0, -4);
}

function find_code () {
    "use strict";
    var codes = document.getElementsByClassName("language-lua");
    for (var i=0; i<codes.length; i++) {
      codes[i].innerHTML = highlight(codes[i].innerHTML);
      codes[i].style.whiteSpace = "";
    }
}
find_code();
